staging/src/k8s.io/client-go/rest/with_retry.go

```diff
@@ -372,6 +379,63 @@ func readAndCloseResponseBody(resp *http.Response) {
 	}
 }
 
+type ResponseCloser struct {
+	once     sync.Once
+	resp     *http.Response
+	respCopy io.ReadWriteCloser
+	err      error
+}
+
+func NewResponseCloser(resp *http.Response) *ResponseCloser {
+	return &ResponseCloser{
+		resp:     resp,
+		respCopy: &buffer{b: &bytes.Buffer{}},
+	}
+}
+
+func (r *ResponseCloser) DrainAndClose() error {
+	r.once.Do(func() {
+		if r.resp == nil || r.resp.Body == nil {
+			return
+		}
+
+		// Ensure the response body is fully read and closed
+		// before we reconnect, so that we reuse the same TCP
+		// connection.
+		const maxBodySlurpSize = 2 << 10
+		defer func() {
+			r.err = r.resp.Body.Close()
+		}()
+
+		if r.resp.ContentLength <= maxBodySlurpSize {
+			var reader io.Reader = &io.LimitedReader{R: r.resp.Body, N: maxBodySlurpSize}
+			reader = io.TeeReader(reader, r.respCopy)
+			io.Copy(ioutil.Discard, reader)
+		}
+	})
+
+	return r.err
+}
+
+type buffer struct {
```

Comment:
can use `io.NopCloser(&bytes.Buffer{})` to do this

Reply:
yeah i had done this initially, but we need a `Write()` method to satisfy `io.Writer` for use in `io.TeeReader`

Reply:
let me know when this is pushed, dropping the `type buffer struct`

Reply:
done @liggitt 
thanks